lib/Utils/Layout/Codegen.cpp

```diff
@@ -123,5 +146,155 @@ FailureOr<std::string> generateLoopNestAsCStr(
+        bodyBuilder) {
+  isl_ast_node* tree = constructAst(rel, ctx_);
+  if (!tree) {
+    return failure();
+  }
+
+  isl_ast_node* node = tree;
+  ValueRange currentIterArgs = initArgs;
+  Location currentLoc = builder_.getUnknownLoc();
+  std::map<std::string, Value> ivToValue;
+  SmallVector<scf::ForOp> loops;
+  // Create the loop nest structure
+  while (isl_ast_node_get_type(node) == isl_ast_node_for) {
+    // Collect the string names of the iterators
+    isl_ast_expr* iter = isl_ast_node_for_get_iterator(node);
+    isl_id* identifier = isl_ast_expr_get_id(iter);
+    std::string name_str(isl_id_get_name(identifier));
+    isl_id_free(identifier);
+    isl_ast_expr_free(iter);
+
+    isl_ast_expr* init = isl_ast_node_for_get_init(node);
+    Value lbInt = buildIslExpr(init, ivToValue, builder_);
+    isl_ast_expr_free(init);
+
+    // The upper bound is always a less than equal operation in the AST codegen.
+    isl_ast_expr* cond = isl_ast_node_for_get_cond(node);
+    assert((isl_ast_expr_get_type(cond) == isl_ast_expr_op &&
+            isl_ast_expr_get_op_type(cond) == isl_ast_op_le) &&
+           "expected upper bound to be a less than equal operation");
+    isl_ast_expr* cond_upper = isl_ast_expr_get_op_arg(cond, 1);
+    Value ubVal = builder_.create<arith::AddIOp>(
+        currentLoc, buildIslExpr(cond_upper, ivToValue, builder_),
+        builder_.create<arith::ConstantIndexOp>(currentLoc, 1));
+    isl_ast_expr_free(cond_upper);
+    isl_ast_expr_free(cond);
+
+    isl_ast_expr* step = isl_ast_node_for_get_inc(node);
+    Value stepInt = buildIslExpr(step, ivToValue, builder_);
+    isl_ast_expr_free(step);
+
+    // Create the scf for loop
+    auto loop = scf::ForOp::create(
+        builder_, currentLoc, lbInt, ubVal, stepInt, currentIterArgs,
+        [&](OpBuilder& nestedBuilder, Location nestedLoc, Value iv,
+            ValueRange args) {
+          ivToValue.insert(std::make_pair(name_str, iv));
+          // It is safe to store ValueRange args because it points to block
+          // arguments of a loop operation that we also own.
+          currentIterArgs = args;
+          currentLoc = nestedLoc;
+        });
+
+    // Set the builder to point to the body of the newly created loop. We don't
+    // do this in the callback because the builder is reset when the callback
+    // returns.
+    builder_.setInsertionPointToStart(loop.getBody());
+    loops.push_back(loop);
+
+    node = isl_ast_node_for_get_body(node);
+  }
+
+  // For all loops but the innermost, yield the results of the nested loop.
+  for (unsigned i = 0, e = loops.size() - 1; i < e; ++i) {
+    builder_.setInsertionPointToEnd(loops[i].getBody());
+    scf::YieldOp::create(builder_, currentLoc, loops[i + 1].getResults());
+  }
+
+  // Generate the indexing expressions.
+  builder_.setInsertionPointToStart(loops.back().getBody());
+  isl_ast_expr* expr = isl_ast_node_user_get_expr(node);
```

Comment:
We should add a check here that expr is not null, since the statement inside the inner most loop can also be an if (or whatever `isl_ast_node_block` is). Maybe just `assert(expr != null && "expected pure noop lest with only user statement")` to get this otherwise primo PR in :)

Reply:
added that + a check to ensure the node the expr came from was a block. if blocks will emit an error with a TODO